### ISSUE TITLE
Import USFM Data

### ISSUE THREAD

**Issue by RRohrssen** [unfoldingWord-dev/ts-requirements#42]
We need to flesh out how this will work for Android and Desktop. The spec is at https://github.com/unfoldingWord-dev/ts-requirements/wiki/Importing-a-Target-Translation#usfm-file

**Comment by RRohrssen** [comment]
Add definition around chunking in the requirements. When importing, verses need to be put into the correct chunk.

**Comment by PhotoNomad0** [comment]
@RRohrssen @jag3773 @neutrinog - the requirements seems a bit loose for USFM import.  What I gather from the document:
- It looks like the requirements here are talking about bible documents only.
- It looks like this USFM document could be a file on an SD card.  What other import methods will we have.  
- It looks like in the case of importing a bible translation we would have to be online to access the chunks.json at https://api.unfoldingword.org/bible/txt/1/

**Comment by PhotoNomad0** [comment]
@RRohrssen @jag3773 @neutrinog - If I took and zipped up the files in door43.org under ./d43-en/ulb/v2/mrk would this be a valid import test case?  I guess we would make no assumption about the contents of a file based on the name, but would just open up each file and try to parse as USFM.

**Comment by da1nerd** [comment]
@PhotoNomad0 USFM currently only applies to Bible projects, but that's not a strict requirement so it shouldn't affect how the import works.
- A single USFM file should represent a complete but single target translation e.g. the _whole_ book of Mark.
- A USFM file _should_ have sufficient information to indicate which project is being translated. However, if it is unclear the app should ask the user what project is represented by the USFM.
- A USFM file generally does not have sufficient information to accurately identify the target language. Therefore the app should ask the user what target language is represented by the USFM.
- A USFM file should be split into chunks (these are the frames/chunks represented in the app) by the `\s5` marker. Please note: when preparing frame id's for use in the app please pad the id's with a `0` e.g. `01`. Also take note of how Psalms is padded. I think it pads to three digits e.g. `001`
- Chapters should be identifiable by the chapter marker `\c` and should be used to correctly organize the split frames.
- The book title should be used for the project title in the app.

Note: that technically USFM should not expect Bible projects exclusively, however as you can see in the above list there is some Bible specific conditions. This is unfortunate, and hopefully we can improve this to be more generic later.

**Comment by PhotoNomad0** [comment]
@neutrinog - from this I would infer that:
- the USFM files for mark (door43.org under ./d43-en/ulb/v2/mrk) are not a good test case since they do not have the \s5 marker.
- I do not have to read the chunks.json at https://api.unfoldingword.org/bible/txt/1/

**Comment by da1nerd** [comment]
Oh. I totally forgot about the chunks.json. no we do use that to split the
chunks. My bad.

**Comment by PhotoNomad0** [comment]
@neutrinog - which means we need to either be online for the import to read the chunk.json file
or we have to store the chunk.json files in the app.

**Comment by da1nerd** [comment]
correct. We will store that in the app. So don't worry about downloading it for now.

**Comment by da1nerd** [comment]
actually... the chunks.json should probably make it's way into the catalog at some point which means it'll get indexed along with everything else. I'll start a new issue for adding this to the indexing and assign it to myself since I need to make sure it works correctly for desktop as well. Afterwards there will be some method somewhere you can call to get the chunks info.

**Comment by da1nerd** [comment]
blocked by #75 

> UPDATE: no longer blocked.

**Comment by da1nerd** [comment]
@PhotoNomad0 the chunks have been added to the index. Please refer to the notes in pull request https://github.com/unfoldingWord-dev/ts-android/pull/1430 to start using the chunks.

**Comment by da1nerd** [comment]
@PhotoNomad0 @jeremymlane when importing USFM we should strip out all `\s5` tags. @jeremymlane I've added the chunks to the index and will add a command shortly so you can retrieve those.

**Comment by da1nerd** [comment]
@PhotoNomad0 let's maybe not support importing a zipped archive of usfm files yet. It introduces complications in the ui that haven't been ironed out yet.

**Comment by cdwhitfield73** [comment]
This is complete in Desktop.

**Comment by klappy** [comment]
Closing for now. Will address improved support with other issues as needed.
